### PR TITLE
gRPC server plugin

### PR DESCRIPTION
**Pull request by tokkee**
This new plugin embeds a gRPC server (http://grpc.io) into collectd. For now, it supports dispatching and querying values.

### CODE REVIEW

File: src/Makefile.am [hunk 1]
```diff
@@ -17,6 +17,12 @@ endif
 AM_CPPFLAGS += -DPLUGINDIR='"${pkglibdir}"'
 AM_CPPFLAGS += -DPKGDATADIR='"${pkgdatadir}"'
 
+AUTOMAKE_OPTIONS = subdir-objects
```
Comment (rubenk) [inline src/Makefile.am]: As of 7cdb614 this is set by configure.ac for the whole project.

File: src/Makefile.am [hunk 2]
```diff
@@ -407,6 +431,17 @@ gmond_la_LDFLAGS = $(PLUGIN_LDFLAGS) $(GANGLIA_LDFLAGS)
 gmond_la_LIBADD = $(GANGLIA_LIBS)
 endif
 
+if BUILD_PLUGIN_GRPC
+pkglib_LTLIBRARIES += grpc.la
+grpc_la_SOURCES = grpc.cc
+nodist_grpc_la_SOURCES = collectd.grpc.pb.cc collectd.pb.cc types.pb.cc
+grpc_la_CPPFLAGS = $(AM_CPPFLAGS) -std=c++11
+grpc_la_CFLAGS = $(AM_CFLAGS)
```
Comment (rubenk) [inline src/Makefile.am]: No need to set this, AM_CFLAGS is prepended by default unless you specifically set per-target CFLAGS.

File: src/collectd.conf.pod [hunk 3]
```diff
@@ -2497,6 +2497,51 @@ source, this is optional. Otherwise the option is required.
 
 =back
 
+=head2 Plugin C<grpc>
+
+The I<grpc> plugin provides an RPC interface to submit values to or query
+values from collectd based on the open source gRPC framework. It exposes and
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/and/an/

File: src/collectd.conf.pod [hunk 4]
```diff
@@ -2497,6 +2497,51 @@ source, this is optional. Otherwise the option is required.
...
+end-point for dispatching values to the daemon.
+
+The B<gRPC> homepage can be found at L<https://grpc.io/>.
+
+=over 4
+
+=item B<Listen> I<Host> I<Port>
+
+The B<Listen> statement sets the network address to bind to. When multiple
+statements are specified, the daemon will bin to all of them. If none are
```
Comment (rubenk) [inline src/collectd.conf.pod]: s/bin/bind/

File: src/Makefile.am [hunk 5]
```diff
@@ -407,6 +431,17 @@ gmond_la_LDFLAGS = $(PLUGIN_LDFLAGS) $(GANGLIA_LDFLAGS)
...
+grpc_la_CXXFLAGS = $(AM_CXXFLAGS) -std=c++11
+grpc_la_LDFLAGS = $(PLUGIN_LDFLAGS)
+grpc_la_LIBADD = -lgrpc++ -lgrpc -lgpr -lprotobuf -lpthread -ldl
```
Comment (rubenk) [inline src/Makefile.am]: Is linking with libpthread and libdl really needed?

File: README [hunk 6]
```diff
@@ -765,6 +772,11 @@ Prerequisites
     The PostgreSQL C client library used by the `postgresql' plugin.
     <http://www.postgresql.org/>
 
+  * libprotobuf, protoc 3.0+ (optional)
+    Used by the `grpc' plugin to generate service stubs and code to handle
+    network packets of collectd's protobuf-based network protocol.
+    <https://developers.google.com/protocol-buffers/>
+
```
Comment (rubenk) [inline README]: Perhaps mention somewhere in here that the grpc plugin needs a C++ compiler that supports C++11?

**Comment by rubenk** [comment]
I left a few comments inline. Otherwise LGTM, nice work!

File: src/Makefile.am (hunk 5 above)
Comment (tokkee) [inline src/Makefile.am]: It's needed by gRPC in general but not for the plugin specifically (similar to your pthread cleanup). Dropped.

File: README (hunk 6 above)
Comment (tokkee) [inline README]: Good idea. Added to the libgrpc entry (which is what requires it).

**Comment by tokkee** [comment]
Thanks for the review. Comments addressed.

**Comment by rubenk** [comment]
Alright, there's a tiny bit of fallout in Jenkins because of this merge.
The rpm builds should be fixed by c089df0.

I'm now searching for our debian packaging scripts to fix those.

I could use a bit of help with the failure of the scan-build job, https://ci.collectd.org/job/master-build-on-xenial-amd64-with-scan-build/lastFailedBuild/consoleFull. I suspect scan-build has trouble deciding if the project is C or C++.

**Comment by rubenk** [comment]
I've never build a Debian package before, but I think https://github.com/collectd/pkg-debian/pull/8 should do the job.

**Comment by tokkee** [comment]
Thanks for the fixes. I failed to reproduce the scan-build failure (using clang 3.5 and 3.8) on Debian Sid. What I found so far does not help really :-/ It shouldn't care much whether the project is C or C++ or a mix, though.

**Comment by rubenk** [comment]
I can't repro it on my Xenial vm. Here's my output:

```
checking for _FILE_OFFSET_BITS value needed for large files... no
checking for gcc... (cached) /usr/share/clang/scan-build-3.8/bin/../libexec/ccc-analyzer
checking whether we are using the GNU C compiler... (cached) yes
checking whether /usr/share/clang/scan-build-3.8/bin/../libexec/ccc-analyzer accepts -g... (cached) yes
checking for /usr/share/clang/scan-build-3.8/bin/../libexec/ccc-analyzer option to accept ISO C89... (cached) none needed
checking whether /usr/share/clang/scan-build-3.8/bin/../libexec/ccc-analyzer understands -c and -o together... (cached) yes
checking whether we are using the GNU C++ compiler... yes
checking whether /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer accepts -g... yes
checking how to run the C++ preprocessor... /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer -E
checking for ld used by /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer... /usr/bin/ld -m elf_x86_64
checking if the linker (/usr/bin/ld -m elf_x86_64) is GNU ld... yes
checking whether the /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking for /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer option to produce PIC... -fPIC -DPIC
checking if /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer PIC flag -fPIC -DPIC works... yes
checking if /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer static flag -static works... yes
checking if /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer supports -c -o file.o... yes
checking if /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer supports -c -o file.o... (cached) yes
checking whether the /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer linker (/usr/bin/ld -m elf_x86_64) supports shared libraries... yes
checking dynamic linker characteristics... (cached) GNU/Linux ld.so
checking how to hardcode library paths into programs... immediate
checking dependency style of /usr/share/clang/scan-build-3.8/bin/../libexec/c++-analyzer... none
checking how to run the C preprocessor... /usr/share/clang/scan-build-3.8/bin/../libexec/ccc-analyzer -E
```

Note the `checking whether we are using the GNU C++ compiler... yes` versus
`checking whether we are using the GNU C++ compiler... no` from the scan-build jenkins job.

**Pull request merged**